File: src/agent/vision/parser.rs
use crate::core::{AgeGroup, InferenceSignal};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct VisionResult {
    pub gender_male: f64,
    pub gender_female: f64,
    pub age_group: Option<String>,
    pub age_confidence: f64,
    pub is_human: bool,
    pub reasoning: String,
}

pub fn strip_markdown(content: &str) -> &str {
    content
        .trim()
        .trim_start_matches("```json")
        .trim_start_matches("```")
        .trim_end_matches("```")
        .trim()
}

pub fn parse_age_group(s: &str) -> Option<AgeGroup> {
    match s {
        "under_18" => Some(AgeGroup::Under18),
        "18-24" => Some(AgeGroup::Age18_24),
        "25-34" => Some(AgeGroup::Age25_34),
        "35-44" => Some(AgeGroup::Age35_44),
        "45-54" => Some(AgeGroup::Age45_54),
        "55-64" => Some(AgeGroup::Age55_64),
        "65+" => Some(AgeGroup::Age65Plus),
        _ => None,
    }
}

pub fn apply_result(signal: &mut InferenceSignal, result: VisionResult) {
    if result.is_human {
        signal.gender_male = result.gender_male;
        signal.gender_female = result.gender_female;
        signal.gender_other = 1.0 - result.gender_male - result.gender_female;

        if let Some(age_str) = &result.age_group {
            if let Some(age_group) = parse_age_group(age_str) {
                signal.set_age_probs(age_group.to_one_hot());
            }
        }

        return signal.reasoning.push(result.reasoning);
    }

    signal
        .reasoning
        .push("Profile picture is not human.".to_string())
}

File: aura-python/aura/agents/vision/parser.py
"""Response parsing for vision analysis."""

from __future__ import annotations

import json
from typing import Any

from aura.core.models import InferenceSignal
from aura.core.types import AgeGroup


def strip_markdown(content: str) -> str:
    """Strip markdown code block markers from content."""
    return (
        content.strip()
        .removeprefix("```json")
        .removeprefix("```")
        .removesuffix("```")
        .strip()
    )


def parse_age_group(s: str) -> AgeGroup | None:
    """Parse age group string to enum."""
    mapping = {
        "under_18": AgeGroup.UNDER_18,
        "18-24": AgeGroup.AGE_18_24,
        "25-34": AgeGroup.AGE_25_34,
        "35-44": AgeGroup.AGE_35_44,
        "45-54": AgeGroup.AGE_45_54,
        "55-64": AgeGroup.AGE_55_64,
        "65+": AgeGroup.AGE_65_PLUS,
    }
    return mapping.get(s)


def parse_response(signal: InferenceSignal, content: str) -> None:
    """Parse the LLM response and update the signal.

    Args:
        signal: The signal to update.
        content: The raw content from the LLM response.
    """
    try:
        result: dict[str, Any] = json.loads(content)
    except json.JSONDecodeError as e:
        signal.reasoning.append(f"Parse error: {e} - Raw: {content}")
        return

    apply_result(signal, result)


def apply_result(signal: InferenceSignal, result: dict[str, Any]) -> None:
    """Apply parsed result to the signal.

    Args:
        signal: The signal to update.
        result: The parsed JSON result.
    """
    is_human = result.get("is_human", False)

    if not is_human:
        signal.reasoning.append("Profile picture is not human.")
        return

    gender_male = result.get("gender_male", 0.0)
    gender_female = result.get("gender_female", 0.0)

    if gender_male > 0.0 or gender_female > 0.0:
        signal.gender_male = gender_male
        signal.gender_female = gender_female
        signal.gender_other = max(0.0, 1.0 - gender_male - gender_female)

    age_group_str = result.get("age_group")
    if age_group_str:
        age_group = parse_age_group(age_group_str)
        if age_group:
            signal.set_age_probs(age_group.to_one_hot())

    reasoning = result.get("reasoning", "")
    if reasoning:
        signal.reasoning.append(reasoning)
